Electron/src/RAM/Autonomy/components/StateDiagram.tsx

```diff
@@ -0,0 +1,186 @@
+import React, { Component } from "react"
+import CSS from "csstype"
+import { rovecomm } from "../../../Core/RoveProtocol/Rovecomm"
+
+const container: CSS.Properties = {
+  display: "flex",
+  flexDirection: "column",
+  fontFamily: "arial",
+  borderTopWidth: "28px",
+  borderColor: "#990000",
+  borderBottomWidth: "2px",
+  borderStyle: "solid",
+  padding: "5px",
+  alignItems: "center",
+  height: "calc(100% - 50px)",
+}
+const label: CSS.Properties = {
+  marginTop: "-10px",
+  position: "relative",
+  top: "24px",
+  left: "3px",
+  fontFamily: "arial",
+  fontSize: "16px",
+  zIndex: 1,
+  color: "white",
+}
+
+interface IProps {
+  style?: CSS.Properties
+}
+
+interface IState {}
+
+const stateEnum = {
+  Idle: 0,
+  Navigating: 1,
+  SearchPattern: 2,
+  ApproachingMarker: 3,
+}
+
+class StateDiagram extends Component<IProps, IState> {
+  canvasRef: any
+
+  constructor(props: IProps) {
+    super(props)
+    this.state = {}
+
+    this.canvasRef = React.createRef()
+    rovecomm.on("CurrentState", (data: any) => this.updateStateDiagram(data))
+  }
+
+  componentDidMount() {
+    this.updateStateDiagram([-1])
+  }
+
+  updateStateDiagram(data: any): void {
+    let text
+
+    const canvas = this.canvasRef.current
+    const context = canvas.getContext("2d")
+    context.clearRect(0, 0, canvas.width, canvas.height)
+
+    const centerW = context.canvas.width / 2
+
+    context.font = "12px Arial"
+
+    text = "Idle"
+    context.textBaseline = "middle"
+    context.textAlign = "right"
+    context.fillStyle = "black"
+    context.fillText(text, centerW - 25, 50)
+    context.beginPath()
+    context.rect(centerW - 20, 30, 40, 40)
+    context.fillStyle = data[0] === stateEnum[text] ? "green" : "white"
+    context.fill()
+    context.stroke()
+
+    context.beginPath()
+    context.moveTo(centerW, 70)
+    context.lineTo(centerW, 110)
+    context.stroke()
+
+    text = "Navigating"
+    context.textBaseline = "middle"
+    context.textAlign = "right"
+    context.fillStyle = "black"
+    context.fillText(text, centerW - 25, 130)
+    context.rect(centerW - 20, 110, 40, 40)
```

Comment:
this might be dumb but what is this the center of?

Reply:
centerW is a constant set on line 66, whose value is the canvas / 2. The canvas should extend to be the size of the entire div. So this is the center of the space that we are trying to draw in.